src/controllers/GovernanceTimelock.sol

```diff
@@ -0,0 +1,158 @@
+
+/// libraries
+import "../libraries/Error.sol";
+
+contract GovernanceTimelock is IGovernanceTimelock {
+    /*///////////////////////////////////////////////////////////////
+                             CONSTANTS
+    //////////////////////////////////////////////////////////////*/
+    uint256 public constant MINIMUM_DELAY = 2 days;
+    uint256 public constant MAXIMUM_DELAY = 30 days;
+
+    /// @notice The time window within which a transaction can be executed, following its ETA. Beyond this point a transaction is considered stale and cannot be executed.
+    uint256 public constant GRACE_PERIOD = 14 days;
+
+    /*/////////////////////////////////////////////////////////////////
+                            STATE VARIABLES
+    ////////////////////////////////////////////////////////////////*/
+    uint256 public txCounter;
+    uint256 public delay = MINIMUM_DELAY;
+
+    /// @notice the admin is the one allowed to schedule events
+    address public admin;
+
+    mapping(uint256 txId => bytes32) public scheduledTransaction;
+    mapping(uint256 txId => bool) public isExecuted;
+
+    /*/////////////////////////////////////////////////////////////////
+                                MODIFIERS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @notice A modifier used for restricting the caller of some functions to be this contract itself.
+    modifier onlySelf() {
+        if (msg.sender != address(this)) {
+            revert Error.INVALID_SELF_CALLER();
+        }
+        _;
+    }
+
+    /// @notice A modifier used for restricting caller to admin contract
+    modifier onlyAdmin() {
+        if (msg.sender != admin) {
+            revert Error.CALLER_NOT_ADMIN();
+        }
+        _;
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                                CONSTRUCTOR
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @param _admin is the address of admin contract that schedule txs
+    constructor(address _admin) {
+        if (_admin == address(0)) {
+            revert Error.ZERO_ADDRESS_INPUT();
+        }
+
+        admin = _admin;
+        emit AdminUpdated(address(0), _admin);
+    }
+
+    /*/////////////////////////////////////////////////////////////////
+                            EXTERNAL FUNCTIONS
+    ////////////////////////////////////////////////////////////////*/
+
+    /// @inheritdoc IGovernanceTimelock
+    function scheduleTransaction(address _target, uint256 _value, bytes memory _data) external override onlyAdmin {
+        if (_target == address(0)) {
+            revert Error.INVALID_TARGET();
+        }
+
+        /// increment tx counter
+        ++txCounter;
+        uint256 eta = block.timestamp + delay;
+
+        scheduledTransaction[txCounter] = keccak256(abi.encodePacked(_target, _value, _data, eta));
+        emit TransactionScheduled(txCounter, _target, _value, _data, eta);
```

Comment:
As noted above, I think we can use the hash as the identifier here instead of `txCounter`